README.md
![Logo for cayenne](https://raw.githubusercontent.com/Heuro-labs/cayenne/master/docs/images/logo.png)

# cayenne : Python package for stochastic simulations

[![Travis Build Status](https://travis-ci.com/Heuro-labs/cayenne.svg?branch=master)](https://travis-ci.com/Heuro-labs/cayenne)
[![Azure Build Status](https://dev.azure.com/srikiranc/cayenne/_apis/build/status/Heuro-labs.cayenne?branchName=master)](https://dev.azure.com/srikiranc/cayenne/_build)
[![codecov](https://codecov.io/gh/Heuro-labs/cayenne/branch/master/graph/badge.svg)](https://codecov.io/gh/Heuro-labs/cayenne)
[![Updates](https://pyup.io/repos/github/Heuro-labs/cayenne/shield.svg)](https://pyup.io/repos/github/Heuro-labs/cayenne/)
[![Documentation Status](https://readthedocs.org/projects/cayenne/badge/?version=latest)](https://cayenne.readthedocs.io/en/latest/?badge=latest)
[![pypi](https://img.shields.io/pypi/v/cayenne.svg)](https://pypi.python.org/pypi/cayenne)
![License](https://img.shields.io/badge/license-Apache%202-blue.svg)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)
[![Gitter chat](https://badges.gitter.im/gitterHQ/gitter.png)](https://gitter.im/cayenne-package)



## Introduction

`cayenne` is a Python package for stochastic simulations. It offers a simple API to define models, perform stochastic simulations with them and visualize the results in a convenient manner.

Currently under active development in the `develop` branch.

## Install

Install with `pip`:

```bash
$ pip install cayenne
```


## Documentation

  - General: <https://cayenne.readthedocs.io>.
  - Benchmark repository, comparing `cayenne` with other stochastic simulation packages: <https://github.com/Heuro-labs/cayenne-benchmarks>

## Usage

A short summary follows, but a more detailed tutorial can be found [here](https://cayenne.readthedocs.io/en/latest/tutorial.html). You can define a model as a Python string (or a text file, see [docs](https://cayenne.readthedocs.io)). The format of this string is loosely based on the excellent [antimony](https://tellurium.readthedocs.io/en/latest/antimony.html#introduction-basics) library, which is used behind the scenes by `cayenne`.

```python
from cayenne.simulation import Simulation
model_str = """
        const compartment comp1;
        comp1 = 1.0; # volume of compartment

        r1: A => B; k1;
        r2: B => C; k2;

        k1 = 0.11;
        k2 = 0.1;
        chem_flag = false;

        A = 100;
        B = 0;
        C = 0;
    """
sim = Simulation.load_model(model_str, "ModelString")
# Run the simulation
sim.simulate(max_t=40, max_iter=1000, n_rep=10)
sim.plot()
```

![Plot of species A, B and C](https://raw.githubusercontent.com/Heuro-labs/cayenne/master/docs/images/plot_basic.png)


### Change simulation algorithm

You can change the algorithm used to perform the simulation by changing the `algorithm` parameter (one of `"direct"`, `"tau_leaping"` or `"tau_adaptive"`)

```python
sim.simulate(max_t=150, max_iter=1000, n_rep=10, algorithm="tau_leaping")
```

Our [benchmarks](https://github.com/Heuro-labs/cayenne-benchmarks) are summarized [below](#benchmarks), and show `direct` to be a good starting point. `tau_leaping` offers greater speed but needs specification and tuning of the `tau` hyperparameter. The `tau_adaptive` is less accurate and a work in progress.

### Run simulations in parallel
You can run the simulations on multiple cores by specifying the `n_procs` parameter

```python
sim.simulate(max_t=150, max_iter=1000, n_rep=10, n_procs=4)
```

### Accessing simulation results

You can access all the results or the results for a specific list of species

```python
# Get all the results
results = sim.results
# Get results only for one or more species
results.get_species(["A", "C"])
```

You can also access the final states of all the simulation runs by

```python
# Get results at the simulation endpoints
final_times, final_states = results.final
```

Additionally, you can access the state a particular time point of interest $t$. `cayenne` will interpolate the value from nearby time points to give an accurate estimate.

```python
# Get results at timepoint "t"
t = 10.0
states = results.get_state(t) # returns a list of numpy arrays
```

<h2 id="benchmarks"> Benchmarks </h2>

| | direct|	tau_leaping |	tau_adaptive |
--- | --- |--- | --- |
cayenne	| :heavy_check_mark: Most accurate yet	| :heavy_check_mark: Very fast but may need manual tuning|	Less accurate than GillespieSSA's version|
Tellurium | :exclamation: Inaccurate for 2nd order | N/A | N/A |
GillespieSSA | Very slow |:exclamation: Inaccurate for initial zero counts | :exclamation: Inaccurate for initial zero counts
BioSimulator.jl |	:exclamation: Inaccurate interpolation | :exclamation: Inaccurate for initial zero counts | :exclamation: Inaccurate for initial zero counts

## License

Copyright (c) 2018-2020, Dileep Kishore, Srikiran Chandrasekaran. Released under: Apache Software License 2.0

## Credits

- [Cython](https://cython.org/)
- [antimony](https://tellurium.readthedocs.io/en/latest/antimony.html)
- [pytest](https://docs.pytest.org)
- [Cookiecutter](https://github.com/audreyr/cookiecutter)
- [audreyr/cookiecutter-pypackage](https://github.com/audreyr/cookiecutter-pypackage)
- [black](https://github.com/ambv/black)
- Logo made with [logomakr](https://logomakr.com/)


HISTORY.rst
=======
History
=======

1.0.3 (2020-08-22)
------------------

Added
+++++
- ``thinning`` parameter to control sampling in ``Simulation.plot``

1.0.2 (2020-08-08)
------------------

Added
+++++
- Warnings about antimony keyword usage in tutorial, ``ModelIO`` class

Fixed
+++++
- ``setup.py`` setup requirements are automatically installed
- ``antimony`` version incompatibility issue

1.0.0 (2020-07-18)
------------------

Added
+++++
- ``Antimony`` support and ``ModelIO`` class, giving easier entry point to load models
- Support for custom species names in plotting and ``Results``
- Support for automatic ``cpu`` core detection
- New logo for ``cayenne``

Changed
+++++++
- Package name changed from ``pyssa`` to ``cayenne``
- Update docs for new API

0.9.1 (2020-05-23)
------------------
Fix interpolation bug in ``Results.get_state``

Fixed
+++++
- ``Results.get_state`` function now adds an epsilon to time

0.9.0 (2019-12-14)
------------------
Replace ``numba`` implementation with ``Cython`` implementation

Fixed
+++++
- Propensity calculation in all algorithms
- Considerable speed-up in algorithm runtimes
- Remove ``volume`` from ``Simulation.simulate`` parameters

Added
+++++
- ``direct`` algorithm in ``Cython``
- ``tau_leaping`` algorithm in ``Cython``
- ``tau_adaptive`` algorithm in ``Cython`` (experimental)
- ``Cython`` to Azure pipeline
- Accuracy tests from `sbml-test-suite <https://github.com/sbmlteam/sbml-test-suite>`_
- HOR property and tests for it
- Code coverage for ``Cython``
- Algorithms page to the documentation
- Examples page to the documentation

Changed
+++++++
- Remove ``numba`` algorithms
- Remove interpolation for direct algorithm
- ``sim.plot`` now plots ``post`` step curve
- Updated tutorial page of the documentation


0.8.2 (2019-04-20)
------------------

Fixed
+++++
- Initialize ``algorithms`` submodule with ``__init__.py``
- Update ``setup.py`` to allow submodule detection

0.8.0 (2019-04-13)
------------------

Added
+++++
- ``Results.get_states`` method - returns state at time ``t``
- Accuracy tests for all algorithms
- Additional consistency checks for ``X0`` and ``k_det``

Changed
+++++++
- Refactor algorithms into sub module ``algorithms``
- Refactor algorithm independent tests

Fixed
+++++
- Indexing issue in propensity calculation in ``direct`` algorithm
- Indexing issue in propensity calculation in ``tau_leaping`` algorithm
- Address edge case X->2X in ``tau_adaptive`` algorithm

0.7.1 (2019-03-23)
------------------

Changed
+++++++
- Refactor ``tau_adaptive``
- Rename ``direct_naive`` to ``direct``

Fixed
+++++
- SSA part of ``tau_adaptive``
- Bug in linux compatibility of ``tau_adaptive``

0.7.0 (2019-02-02)
------------------

Added
+++++
- Support for the ``tau_adaptive`` algorithm
- Support for multiprocessing

Fixed
+++++
- Transpose stoichiometric matrix
- Update references in docstrings

Changed
+++++++
- Use ``TINY`` and ``HIGH`` for status estimation
- Use ``np.int64`` and ``np.float64`` explicitly

Chore
+++++
- Update dependencies
- Add azure pipelines for testing on Windows

0.6.0 (2018-12-16)
------------------

Added
+++++
- Updated ``direct_naive`` docstring
- Support for the ``tau_leaping`` algorithm
- Species name support for plotting

Fixed
+++++
- Check for sum propensities uses threshold instead of equality
- Add check for type of ``max_iter``

Changed
+++++++
- Update ``roulette_selection`` to use `np.searchsorted`
- Minor changes to ``numpy`` style usage

Chore
+++++
- Add ``codecov``
- Travis pypi autodepolyment
- Parameterize tests with algorithm name
- Add details about ``tau_leaping`` to docs and README


0.5.4 (2018-12-02)
------------------

Added
+++++
- badge to readme

0.5.3 (2018-12-02)
------------------

Added
+++++
- plot to pypi

Changed
+++++++
- fix bumpversion/black issue
- remove history from package long_description


0.5.0 (2018-12-01)
------------------

First public release!!

Added
+++++
- testpypi deployment
- pyup security checking
- readthedocs deployment
- Tutorials and documentation
- Plotting functionality through ``Simulation.plot``

Changed
+++++++
- ``Simulation.results`` is now a property
- Updated tests to support the new api changes

Chore
+++++
- Updated the README


0.4.0 (2018-11-23)
------------------

Added
+++++
- ``Simulation`` class - main class for running simulations
- ``Results`` class - for storing and acessing simulation results
- ``Simulation.simulate`` function that returns an instance of the ``Results`` class

Changed
+++++++
- Refactor ``get_kstoc`` and ``roulette_selection`` into ``utils.py``
- Refactor ``direct_naive`` into ``direct_naive.py``
- Delete ``pyssa.py`` and replace with ``Simulation`` class

Chore
+++++
- Add license and code-style badges
- Use ``black`` for code-formatting


0.2.0 (2018-11-10)
------------------

Added
+++++

- Naive implementation of the Gillepsie algorithm in ``numba``
- Tests - sanity checks, bifurcation and long running simulation
- CI on ``travis``


0.1.0 (2018-08-08)
------------------

* First commit


pyssa/simulation.py
"""
    The main class for running stochastic simulation
"""

import multiprocessing as mp
from functools import partial
from typing import List, Optional
from warnings import warn

import matplotlib.lines as mlines
import matplotlib.pyplot as plt
import numpy as np

from .algorithms.direct import direct
from .algorithms.tau_adaptive import tau_adaptive
from .algorithms.tau_leaping import tau_leaping
from .model_io import ModelIO
from .results import Results


def wrapper(x, func):
    return func(*x)


class Simulation:
    """
        A main class for running simulations.

        Parameters
        ----------
        species_names : List[str]
            List of species names
        rxn_names : List[str]
            List of reaction names
        react_stoic: (ns, nr) ndarray
            A 2D array of the stoichiometric coefficients of the reactants.
            Reactions are columns and species are rows.
        prod_stoic: (ns, nr) ndarray
            A 2D array of the stoichiometric coefficients of the products.
            Reactions are columns and species are rows.
        init_state: (ns,) ndarray
            A 1D array representing the initial state of the system.
        k_det: (nr,) ndarray
            A 1D array representing the deterministic rate constants of the
            system.
        volume: float, optional
            The volume of the reactor vessel which is important for second
            and higher order reactions. Defaults to 1 arbitrary units.
        chem_flag: bool, optional
            If True, divide by Na (Avogadro's constant) while calculating
            stochastic rate constants. Defaults to ``False``.

        Attributes
        ----------
        results: Results
            The results instance

        Raises
        ------
        ValueError
            If supplied with order > 3.

        Examples
        --------
        >>> V_r = np.array([[1,0],[0,1],[0,0]])
        >>> V_p = np.array([[0,0],[1,0],[0,1]])
        >>> X0 = np.array([10,0,0])
        >>> k = np.array([1,1])
        >>> sim = Simulation(V_r, V_p, X0, k)

        Notes
        -----
        Stochastic reaction rates depend on the size of the system for second
        and third order reactions. By this, we mean the volume of the system in
        which the reactants are contained. Intuitively, this makes sense
        considering that collisions between two or more molecules becomes
        rarer as the size of the system increases. A detailed mathematical
        treatment of this idea can be found in [3]_ .

        In practice, this means that ``volume`` and ``chem_flag`` need to be
        supplied for second and third order reactions. ``volume``
        represents the size of the system containing the reactants.

        In chemical systems ``chem_flag`` should generally be set to ``True``
        as ``k_det`` is specified in units of molarity or M or mol/L.
        For example, a second order rate constant could be = 0.15 mol / (L s).
        Then Avogadro's constant (:math:`N_a`) is used for normalization while
        computing ``k_stoc`` (:math:`c_\mu` in [3]_ ) from ``k_det``.

        In biological systems, ``chem_flag`` should be generally be set to
        ``False`` as ``k_det`` is specified in units of copies/L or CFU/L.
        For example, a second order rate constant could be = 0.15 CFU / (L s).

        References
        ----------
        .. [3] Gillespie, D.T., 1976.
            A general method for numerically simulating the stochastic time
            evolution of coupled chemical reactions. J. Comput. Phys. 22, 403–434.
            doi:10.1016/0021-9991(76)90041-3.

    """

    _results: Optional[Results] = None

    def __init__(
        self,
        species_names: List[str],
        rxn_names: List[str],
        react_stoic: np.ndarray,
        prod_stoic: np.ndarray,
        init_state: np.ndarray,
        k_det: np.ndarray,
        chem_flag: bool = False,
        volume: float = 1.0,
    ) -> None:
        self.species_names = species_names
        self.rxn_names = rxn_names
        self._react_stoic = react_stoic
        self._prod_stoic = prod_stoic
        self._init_state = init_state
        self._k_det = k_det
        self._chem_flag = chem_flag
        self._ns = self._react_stoic.shape[0]
        self._nr = self._react_stoic.shape[1]
        self._volume = volume
        self._orders = np.sum(
            self._react_stoic, axis=0
        )  # Order of rxn = number of reactants
        self._check_consistency()

    def _check_consistency(self):
        if len(self.species_names) != self._ns:
            raise ValueError(
                "Species names must match the stoichiometric matrix shapes."
            )
        if len(self.rxn_names) != self._nr:
            raise ValueError(
                "Reaction names must match the stoichiometric matrix shapes."
            )
        if (self._ns != self._prod_stoic.shape[0]) or (
            self._nr != self._prod_stoic.shape[1]
        ):
            raise ValueError("react_stoic and prod_stoic should be the same shape.")
        if np.any(self._react_stoic < 0):
            raise ValueError("react_stoic cannot have negative elements.")
        if np.any(self._prod_stoic < 0):
            raise ValueError("V_p cannot have negative elements.")
        if np.any(self._init_state < 0):
            raise ValueError("Initial numbers in X0 can't be negative.")
        if np.any(self._k_det < 0):
            raise ValueError("Rate constant(s) can't be negative.")
        if self._k_det.shape[0] != self._nr:
            raise ValueError("Number of rate constants must equal number of reactions")
        if len(self._k_det.shape) != 1:
            raise ValueError("k_det must be a 1-D array")
        if self._chem_flag not in (True, False):
            raise ValueError("chem_flag must be a boolean True or False.")
        if np.max(self._orders) > 3:
            raise ValueError("Order greater than 3 not suppported.")
        if self._ns != self._init_state.shape[0]:
            raise ValueError(
                "X0 must have be of length = num. of species (or rows of V_r)"
            )
        if len(self._init_state.shape) != 1:
            raise ValueError("X0 must be a 1-D array")

    @property
    def results(self) -> Optional[Results]:
        """
            The ``Results`` instance of the simulation

            Returns
            -------
            Optional[Results]
        """
        if self._results is None:
            warn(
                "Run `Simulation.simulate` before requesting the results object",
                Warning,
            )
            return self._results
        else:
            return self._results

    @classmethod
    def load_model(cls, contents: str, contents_type: str) -> "Simulation":
        """
            Load model contents into a Simulation object

            Parameters
            ----------
            model_contents : str
                Either the model string or the file path
            content_type : str, {"ModelString", "ModelFile"}
                The type of the model

            Returns
            -------
            sim : Simulation
                An instance of the Simulation class.
         """
        modelio = ModelIO(contents, contents_type)
        (
            species_names,
            rxn_names,
            react_stoic,
            prod_stoic,
            init_state,
            k_det,
            chem_flag,
            volume,
        ) = modelio.args
        return cls(
            species_names,
            rxn_names,
            react_stoic,
            prod_stoic,
            init_state,
            k_det,
            chem_flag,
            volume,
        )

    def simulate(
        self,
        max_t: float = 10.0,
        max_iter: int = 1000,
        seed: int = 0,
        n_rep: int = 1,
        n_procs: Optional[int] = 1,
        algorithm: str = "direct",
        debug: bool = False,
        **kwargs,
    ) -> None:
        """
            Run the simulation

            Parameters
            ----------
            max_t: float, optional
                The end time of the simulation.
                The default is 10.0.
            max_iter: int, optional
                The maximum number of iterations of the simulation loop.
                The default is 1000 iterations.
            seed: int, optional
                The seed used to generate simulation seeds.
                The default value is 0.
            n_rep: int, optional
                The number of repetitions of the simulation required.
                The default value is 1.
            n_procs: int, optional
                The number of cpu cores to use for the simulation.
                Use ``None`` to automatically detect number of cpu cores.
                The default value is 1.
            algorithm: str, optional
                The algorithm to be used to run the simulation.
                The default value is ``"direct"``.

            Notes
            -----
            The status indicates the status of the simulation at exit. Each
            repetition will have a status associated with it, and these are
            accessible through the ``Simulation.results.status_list``.

            status: int
                Indicates the status of the simulation at exit.

                1: Succesful completion, terminated when ``max_iter`` iterations reached.

                2: Succesful completion, terminated when ``max_t`` crossed.

                3: Succesful completion, terminated when all species went extinct.

                -1: Failure, order greater than 3 detected.

                -2: Failure, propensity zero without extinction.
        """
        tlist = []
        xlist = []
        status_list = []

        if not isinstance(seed, int):
            raise TypeError("Seed should be of type int")
        np.random.seed(seed)
        sim_seeds = np.random.randint(low=0, high=1e7, size=n_rep)

        if not isinstance(max_iter, int):
            raise TypeError("max_iter should be of type int")

        algo_args = []
        if algorithm == "direct":
            for index in range(n_rep):
                algo_args.append(
                    (
                        self._react_stoic,
                        self._prod_stoic,
                        self._init_state,
                        self._k_det,
                        max_t,
                        max_iter,
                        self._volume,
                        sim_seeds[index],
                        self._chem_flag,
                    )
                )
                algo = direct
        elif algorithm == "tau_leaping":
            if "tau" in kwargs.keys():
                tau = kwargs["tau"]
            else:
                tau = 0.1
            for index in range(n_rep):
                algo_args.append(
                    (
                        self._react_stoic,
                        self._prod_stoic,
                        self._init_state,
                        self._k_det,
                        tau,
                        max_t,
                        self._volume,
                        sim_seeds[index],
                        self._chem_flag,
                    )
                )
                algo = tau_leaping
        elif algorithm == "tau_adaptive":
            if "epsilon" in kwargs.keys():
                epsilon = kwargs["epsilon"]
            else:
                epsilon = 0.03
            if "nc" in kwargs.keys():
                nc = kwargs["nc"]
            else:
                nc = 10
            hor = self.HOR
            for index in range(n_rep):
                algo_args.append(
                    (
                        self._react_stoic,
                        self._prod_stoic,
                        self._init_state,
                        self._k_det,
                        hor.astype(np.int),
                        nc,
                        epsilon,
                        max_t,
                        max_iter,
                        self._volume,
                        sim_seeds[index],
                        self._chem_flag,
                    )
                )
            algo = tau_adaptive
        else:
            raise ValueError("Requested algorithm not supported")
        algo_func = partial(wrapper, func=algo)
        if debug:
            results = map(algo_func, algo_args)
        else:
            with mp.Pool(processes=n_procs) as pool:
                results = pool.map(algo_func, algo_args)
        for t, X, status in results:
            tlist.append(t)
            xlist.append(X)
            status_list.append(status)
        self._results = Results(
            self.species_names,
            self.rxn_names,
            tlist,
            xlist,
            status_list,
            algorithm,
            sim_seeds,
        )

    @property
    def HOR(self) -> np.ndarray:
        """
            Determine the HOR vector. HOR(i) is the highest order of reaction
            in which species S_i appears as a reactant.

            Returns
            -------
            HOR: np.ndarray
                Highest order of the reaction for the reactive species as
                defined under Eqn. (27) of [1]_. HOR can be 1, 2 or 3
                if the species appears only once in the reactants.
                If HOR is -2, it appears twice in a second order reaction.
                If HOR is -3, it appears thrice in a third order reaction.
                If HOR is -32, it appears twice in a third order reaction.
                The corresponding value of ``g_i`` in Eqn. (27) is handled
                by ``tau_adaptive``.

            References
            ----------
            .. [1] Cao, Y., Gillespie, D.T., Petzold, L.R., 2006.
                Efficient step size selection for the tau-leaping simulation
                method. J. Chem. Phys. 124, 044109. doi:10.1063/1.2159468
        """
        ns = self._react_stoic.shape[0]
        HOR = np.zeros(ns, dtype=np.int32)
        orders = np.sum(self._react_stoic, axis=0)
        for ind in range(ns):
            this_orders = orders[np.where(self._react_stoic[ind, :] > 0)[0]]
            if len(this_orders) == 0:
                HOR[ind] = 0
                continue
            HOR[ind] = np.max(this_orders)
            if HOR[ind] == 1:
                continue
            order_2_indices = np.where(orders == 2)
            this_react_stoic = self._react_stoic[ind, :]
            if order_2_indices[0].size > 0:
                if np.max(this_react_stoic[order_2_indices[0]]) == 2 and HOR[ind] == 2:
                    HOR[ind] = -2  # g_i should be (2 + 1/(x_i-1))
            if np.where(orders == 3):
                if (
                    HOR[ind] == 3
                    and np.max(this_react_stoic[np.where(this_orders == 3)[0]]) == 2
                ):
                    HOR[ind] = -32  # g_i should be (3/2 * (2 + 1/(x_i-1)))
                elif (
                    HOR[ind] == 3
                    and np.max(this_react_stoic[np.where(this_orders == 3)[0]]) == 3
                ):
                    HOR[ind] = -3  # g_i should be(3 + 1/(x_i-1) + 2/(x_i-2))
        return HOR

    def plot(self, species_names: list = None, new_names: list = None):
        """
            Plot the simulation

            Parameters
            ----------
            species_names: list, optional
                The names of the species to be plotted (``list`` of ``str``).
                The default is ``None`` and  plots all species.
            new_names: list, optional
                The names of the species to be plotted.
                The default is ``"xi"`` for species ``i``.

            Returns
            -------
            fig: class 'matplotlib.figure.Figure'
                Figure object of the generated plot.
            ax: class 'matplotlib.axes._subplots.AxesSubplot
                Axis objected of the generated plot.
        """

        if self._results is None:
            raise ValueError("Simulate not run.")
        else:
            if species_names is None:
                species_names = self.species_names
            n_species = len(species_names)
            prop_cycle = plt.rcParams["axes.prop_cycle"]
            colors = prop_cycle.by_key()["color"]
            fig, ax = plt.subplots()
            res = self._results
            legend_handlers = [0] * n_species
            generic_names = [""] * n_species
            for index1 in range(n_species):
                legend_handlers[index1] = mlines.Line2D([], [], color=colors[index1])
                this_species = species_names[index1]
                generic_names[index1] = this_species
                n_reps = len(res)
                for index2 in range(n_reps):
                    ax.step(
                        res.t_list[index2],
                        res.get_species([this_species])[index2],
                        color=colors[index1],
                        where="post",
                    )
            if new_names is None:
                new_names = self.species_names
            fig.legend(legend_handlers, new_names)
            return fig, ax
